p11-kit/generate-keypair.c

```diff
@@ -0,0 +1,467 @@
+
+	return false;
+}
+
+static int
+generate_keypair (const char *token_str,
+		  const char *label,
+		  CK_MECHANISM *mechanism,
+		  CK_ULONG bits,
+		  const p11_buffer *ec_params)
+{
+	int ret = 1;
+	CK_RV rv;
+	const char *pin = NULL;
+	P11KitUri *uri = NULL;
+	P11KitIter *iter = NULL;
+	CK_FUNCTION_LIST **modules = NULL;
+	CK_FUNCTION_LIST *module = NULL;
+	CK_SESSION_HANDLE session = 0;
+	CK_SLOT_ID slot = 0;
+	CK_ATTRIBUTE *pubkey = NULL, *privkey = NULL;
+	CK_OBJECT_HANDLE pubkey_obj, privkey_obj;
+
+	if (!get_templates (label, mechanism->mechanism, &bits, ec_params, &pubkey, &privkey)) {
+	        p11_message (_("failed to create key templates"));
+		goto cleanup;
+	}
+
+	uri = p11_kit_uri_new ();
+	if (uri == NULL) {
+		p11_message (_("failed to allocate memory for URI"));
+		goto cleanup;
+	}
+
+	if (p11_kit_uri_parse (token_str, P11_KIT_URI_FOR_TOKEN, uri) != P11_KIT_URI_OK) {
+		p11_message (_("failed to parse the token URI"));
+		goto cleanup;
+	}
+
+	modules = p11_kit_modules_load_and_initialize (0);
+	if (modules == NULL) {
+		p11_message (_("failed to load and initialize modules"));
+		goto cleanup;
+	}
+
+	iter = p11_kit_iter_new (uri, P11_KIT_ITER_WITH_TOKENS | P11_KIT_ITER_WITHOUT_OBJECTS);
+	if (iter == NULL) {
+		p11_message (_("failed to initialize iterator"));
+		goto cleanup;
+	}
+
+	p11_kit_iter_begin (iter, modules);
+	rv = p11_kit_iter_next (iter);
+	if (rv != CKR_OK) {
+		p11_message (_("failed to find the token: %s"), p11_kit_strerror (rv));
+		goto cleanup;
+	}
+
+	module = p11_kit_iter_get_module (iter);
+	if (module == NULL) {
+		p11_message (_("failed to obtain module"));
+		goto cleanup;
+	}
+
+	slot = p11_kit_iter_get_slot (iter);
+	if (slot == 0) {
+		p11_message (_("failed to obtain slot"));
+		goto cleanup;
+	}
+
+	rv = module->C_OpenSession (slot, CKF_SERIAL_SESSION | CKF_RW_SESSION, NULL, NULL, &session);
+	if (rv != CKR_OK) {
+		p11_message (_("failed to open session: %s"), p11_kit_strerror (rv));
+		goto cleanup;
+	}
+
+	pin = p11_kit_uri_get_pin_value (uri);
+	if (pin != NULL) {
+		rv = module->C_Login (session, CKU_USER, (unsigned char *)pin, strlen (pin));
+		if (rv != CKR_OK) {
+			p11_message (_("failed to login: %s"), p11_kit_strerror (rv));
+			goto cleanup;
+		}
+	}
+
+	rv = module->C_GenerateKeyPair (session, mechanism,
+					pubkey, p11_attrs_count (pubkey),
+					privkey, p11_attrs_count (privkey),
+					&pubkey_obj, &privkey_obj);
+	if (rv != CKR_OK) {
+		p11_message (_("key-pair generation failed: %s"), p11_kit_strerror (rv));
+		goto cleanup;
+	}
+
+	ret = 0;
+
+cleanup:
+	if (session != 0)
+		module->C_CloseSession (session);
+	p11_attrs_free (pubkey);
+	p11_attrs_free (privkey);
+	p11_kit_iter_free (iter);
+	p11_kit_uri_free (uri);
+	if (modules != NULL)
+		p11_kit_modules_finalize_and_release (modules);
+
+	return ret;
+}
+
+int
+p11_kit_generate_keypair (int argc,
+			  char *argv[])
+{
+	int opt, ret = 2;
+	char *label = NULL;
+	CK_ULONG bits = 0;
+	p11_buffer ec_params;
```

Comment:
`p11_buffer` is used here just because it's a convenient structure with data and size? It's confusing as it gives the reader an impression that this data will be modified afterwards. Let's just use `const uint8_t *` with size.

Reply:
done